[tree] Add missing `override` specifiers in tree/dataframe

guitargeek opened the pull request:
For developers, it is unconvenient that the `override` specifier
that flags member functions as overriding on first sight is not used so
much in `hist`.

I think it is a good time to add the missing `override` specifiers
to more ROOT components now, as I have already done it for `roofit`,
`math`, and `hist`, where it was really a game changer for me because it
made reading und understanding the code easier!

See:
  * https://github.com/root-project/root/pull/10083
  * https://github.com/root-project/root/pull/9808
  * https://github.com/root-project/root/pull/9883

This commit was generated more or less automatically with this Python
script that uses `clang-tidy`:

```Python
import os
import glob
import subprocess
import tqdm

"""
For clang-tidy to work, you have to copy the compile_commands.json from the
build directory back into the repo directory (just like in
.ci/copy_headers.sh).
"""

def get_sources(directory, extensions):

    files = []

    for ext in extensions:
        files += glob.glob(
            os.path.join(directory, "**/*" + ext), recursive=True
        )

    return files

"""
Recursively find extensions in directory, to figure out whic hextensions
should be globbed for.
   find . -type f -name '*.*' | sed 's|.*\.||' | sort -u
"""
extensions = [".h", ".hpp", ".hxx", ".cpp", ".cc", ".cxx"]

"""
Some extensions are recognized as C and not as C++ files by clang-tidy. We
need to rename them, and this dict specifies how file extensions should be
replaced.
"""
rename_dict = {".h": ".hpp"}

files = get_sources("tree", extensions)

cflags = (
    subprocess.check_output(["root-config", "--cflags"]).strip().decode("utf-8")
)

for file in tqdm.tqdm(files):

    file_renamed = file
    for ext, ext_renamed in rename_dict.items():
        if file.endswith(ext):
            file_renamed = file.replace(ext, ext_renamed)

    if file_renamed != file:
        os.rename(file, file_renamed)

    cmd = [
        "clang-tidy",
        "-checks=modernize-use-override",
        "--fix",
        file_renamed,
        "--",
    ] + cflags.split(" ")
    subprocess.call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

    if file_renamed != file:
        os.rename(file_renamed, file)

"""
Finally, replace the ClassDef with the ClassDefOverride macros.
  find tree -type f -print | xargs sed -i 's/ClassDef(/ClassDefOverride(/'
...and change back the ClassDefOverride of non-overriding base classes.
"""
```

eguiraud reviewed (commented):
Hi Jonas, thank you very much for this.

For `tree/dataframe`:
- `lexertk.hpp` should be left alone
- most if not all `override` should actually be `final`
- I would also leave `tree/dataframe/test` alone, but since we already have the changes we might as well keep them 
- this might be subjective, but I am not a fan of adding `override` to all destructors of inheriting classes, it seems redundant

guitargeek commented:
Hi!

> most if not all override should actually be final

Are your referring to `dataframe` here or all the tree subsystem? And would this not break usercode where people are further inheriting from the overriding classes?

eguiraud commented:
> Are your referring to dataframe here or all the tree subsystem?

dataframe

> would this not break usercode where people are further inheriting from the overriding classes?

In RDF, the only public classes that are designed for inheritance are `RActionImpl` and `RDataSource` (they should not be final or have methods marked final, but they do not have a base class either so I don't think this PR touches them). All other classes are in namespace `Detail` or `Internal` and I cannot think of any case where I designed them for multiple levels of inheritance.

guitargeek commented:
Alright, I have reverted the addition of `override` to the destructors in `dataframe` and added `final` where I could. That was indeed almost everywhere, except for `RRange::GetGraph()`.

pcanal commented:
It might be useful to seperate this PR in 3 (RDataFrame, TTree and RNTuple).   On first glance, the set of over-ride in TTree seems to be much lower in number that I would have expected.

guitargeek commented:
Hi @eguiraud, I have implemented your suggestions and split of the non-RDF changes to other PRs.

eguiraud commented on tree/dataframe/inc/ROOT/RDF/RLoopManager.hxx:
```diff
@@ -211,7 +211,7 @@ public:
 
    /// End of recursive chain of calls, does nothing
-   void AddFilterName(std::vector<std::string> &) {}
+   void AddFilterName(std::vector<std::string> &) override {}
```
why can this not be `final`? there should be no type inheriting from `RLoopManager`

eguiraud commented on tree/dataframe/inc/ROOT/RDF/RLoopManager.hxx:
```diff
@@ -223,7 +223,7 @@ public:
 
    std::shared_ptr<ROOT::Internal::RDF::GraphDrawing::GraphNode>
-   GetGraph(std::unordered_map<void *, std::shared_ptr<ROOT::Internal::RDF::GraphDrawing::GraphNode>> &visitedMap);
+   GetGraph(std::unordered_map<void *, std::shared_ptr<ROOT::Internal::RDF::GraphDrawing::GraphNode>> &visitedMap) override;
```
same as above

eguiraud commented on tree/dataframe/inc/ROOT/RDF/RRangeBase.hxx:
```diff
@@ -50,11 +50,11 @@ public:
 
    RRangeBase &operator=(const RRangeBase &) = delete;
-   virtual ~RRangeBase();
+   ~RRangeBase();
```
same comment as per `~RFilterBase()`

eguiraud commented on tree/dataframe/inc/ROOT/RDF/RRangeBase.hxx:
```diff
@@ -50,11 +50,11 @@ public:
    void InitNode() { ResetCounters(); }
-   virtual std::shared_ptr<RDFGraphDrawing::GraphNode>
-   GetGraph(std::unordered_map<void *, std::shared_ptr<RDFGraphDrawing::GraphNode>> &visitedMap) = 0;
+   std::shared_ptr<RDFGraphDrawing::GraphNode>
+   GetGraph(std::unordered_map<void *, std::shared_ptr<RDFGraphDrawing::GraphNode>> &visitedMap) override = 0;
```
```suggestion
```
?

eguiraud commented on tree/dataframe/inc/ROOT/RDF/RRange.hxx:
```diff
@@ -109,9 +109,9 @@ public:
+   void AddFilterName(std::vector<std::string> &filters) final { fPrevNode.AddFilterName(filters); }
    std::shared_ptr<RDFGraphDrawing::GraphNode>
-   GetGraph(std::unordered_map<void *, std::shared_ptr<RDFGraphDrawing::GraphNode>> &visitedMap)
+   GetGraph(std::unordered_map<void *, std::shared_ptr<RDFGraphDrawing::GraphNode>> &visitedMap) override
```
```suggestion
   GetGraph(std::unordered_map<void *, std::shared_ptr<RDFGraphDrawing::GraphNode>> &visitedMap) final
```
?

eguiraud commented on tree/dataframe/inc/ROOT/RDF/RVariedAction.hxx:
```diff
@@ -189,7 +189,7 @@ public:
    }
 
-   [[noreturn]] std::unique_ptr<RActionBase> MakeVariedAction(std::vector<void *> &&)
+   [[noreturn]] std::unique_ptr<RActionBase> MakeVariedAction(std::vector<void *> &&) override
```
```suggestion
   [[noreturn]] std::unique_ptr<RActionBase> MakeVariedAction(std::vector<void *> &&) final
```
?

eguiraud reviewed (commented):
Thank you @guitargeek , a couple of things left to clarify/discuss before approval.

guitargeek commented on tree/dataframe/inc/ROOT/RDF/RFilterBase.hxx:
```diff
@@ -55,7 +55,7 @@ public:
    RFilterBase &operator=(const RFilterBase &) = delete;
 
-   virtual ~RFilterBase();
+   ~RFilterBase();
```
The thing is that this constructor is already overriding the constructor of the base class, so actually `clang-tidy` suggests here:

```C++
~RFilterBase() override;
```

But then you didn't like that (see https://github.com/root-project/root/pull/11285#pullrequestreview-1090458149):

> I am not a fan of adding `override` to all destructors of inheriting classes, it seems redundant

so maybe we should use `override` for the destructors after all, because `virtual` suggests that this class is not overriding a base class, and having no keywords is also bad as you explained now

guitargeek commented:
Hi @eguiraud, thanks for the review and the comments! I addressed almost all of them, except for the ones related to the destructors.

There are three options we have for the overriding destructors, and there are arguments against all of them:

1. Add `override` as suggested by `clang-tidy`, which you don't like because it appears redundant
2. Adding nothing, but then removing the base class would break downsteam inheritance
3. Adding `virtual`, but this implies that it's the first function in the override chain

Let me know what you want me to do! Personally, I'm for doing what clang-tidy wants :)

Clang tidy suggests to add `override` to the constructor if it overrides a base class,

eguiraud commented:
Thank you for your patience @guitargeek :)

Given clang-tidy's behavior, I think you are right, option 1 is the best: I'd rather have clang-tidy-compliant code than code that looks exactly how I like it :)

eguiraud reviewed (approved):
Thanks!
